connmgr/connmanager.go

```diff
@@ -0,0 +1,336 @@
+				"-- retrying connection in: %v", maxFailedAttempts,
+				cm.cfg.RetryDuration)
+			time.AfterFunc(cm.cfg.RetryDuration, func() {
+				cm.Connect(&ConnReq{Addr: cm.cfg.GetNewAddress()})
+			})
+		} else {
+			go cm.Connect(&ConnReq{Addr: cm.cfg.GetNewAddress()})
+		}
+	}
+}
+
+// connHandler handles all connection related requests.  It must be run as a
+// goroutine.
+//
+// The connection handler makes sure that we maintain a pool of active outbound
+// connections so that we remain connected to the network.  Connection requests
+// are processed and mapped by their assigned ids.
+func (cm *ConnManager) connHandler() {
+out:
+	for {
+		select {
+		case req := <-cm.requests:
+			switch msg := req.(type) {
+
+			case handleConnected:
+				connReq := msg.c
+				connReq.updateState(ConnEstablished)
+				connReq.conn = msg.conn
+				cm.conns[connReq.id] = connReq
+				log.Infof("Connection established: %v", connReq.Addr)
+				connReq.retryCount = 0
+				cm.failedAttempts = 0
+
+				if cm.cfg.OnConnection != nil {
+					go cm.cfg.OnConnection(connReq, msg.conn)
+				}
+
+			case handleDisconnected:
+				if connReq, ok := cm.conns[msg.id]; ok {
+					connReq.updateState(ConnDisconnected)
+					if connReq.conn != nil {
+						connReq.conn.Close()
+					}
+					log.Debugf("Connection closed: %v", connReq.Addr)
+					delete(cm.conns, msg.id)
+
+					if cm.cfg.OnDisconnection != nil {
+						go cm.cfg.OnDisconnection(connReq)
+					}
+
+					cm.handleFailedConn(connReq, msg.retry)
+				}
+
+			case handleFailed:
+				connReq := msg.c
+				connReq.updateState(ConnFailed)
+				log.Debugf("Connection failed %v: %v", connReq.Addr, msg.err)
+
+				cm.handleFailedConn(connReq, true)
+			}
+
+		case <-cm.quit:
+			break out
+		}
+	}
+
+	cm.wg.Done()
+	log.Trace("Connection handler done")
+}
+
+// Connect assigns an id and dials a connection to the address of the
+// connection request.
+func (cm *ConnManager) Connect(c *ConnReq) {
+	if atomic.LoadInt32(&cm.stop) != 0 {
+		return
+	}
+	atomic.StoreUint64(&c.id, atomic.AddUint64(&cm.connReqCount, 1))
+	log.Debugf("Attempting connection: %v", c.Addr)
+	conn, err := cm.cfg.Dial("tcp", c.Addr)
+	if err != nil {
+		cm.requests <- handleFailed{c, err}
+	} else {
+		cm.requests <- handleConnected{c, conn}
+	}
+}
+
+// Disconnect disconnects the connection corresponding to the given connection
+// id. If permanent, the connection will be retried with an increasing backoff
+// duration.
+func (cm *ConnManager) Disconnect(id uint64) {
+	if atomic.LoadInt32(&cm.stop) != 0 {
+		return
+	}
+	cm.requests <- handleDisconnected{id, true}
+}
+
+// Remove removes the connection corresponding to the given connection
+// id from known connections.
+func (cm *ConnManager) Remove(id uint64) {
+	if atomic.LoadInt32(&cm.stop) != 0 {
+		return
+	}
+	cm.requests <- handleDisconnected{id, false}
+}
+
+// Start launches the connection manager and begins connecting to the network.
+func (cm *ConnManager) Start() {
```

Comment:
This currently allows the `Start()` method to be called multiple times in a non-idempotent manner, that doesn't seem desirable. 

I think the code block below should be guarded with something like: 
```go
if !atomic.CompareAndSwapUint64(&cm.stop, 0, 1) {
    return
}
```

With this guard, calling the `Start` method multiple times (even concurrently) is idempotent. Alternatively, we could use a `sync.Once` here, though I lean towards the CAS call.